### PR TITLE
chore: remove curl, mkdir dependency

### PR DESCRIPTION
**Pull request by noahnu**
## Description

- Removes dependency on curl & mkdir in node installer, for #348.

### Checklist
- [ ] This PR has updated documentation
- [ ] This PR has sufficient testing

### CODE REVIEW

File: scripts/install.js [hunk 1]
```diff
@@ -67,25 +71,58 @@ function escapeRegExp(src) {
...
 async function ensureDir(dirPath) {
-    if (fs.existsSync(dirPath)) return
-    execSync(`mkdir -p ${dirPath}`)
+    if (!fs.existsSync(dirPath)) {
+        const normalizedPath = path.normalize(path.relative('', dirPath))
+        ;[...normalizedPath.split(path.sep), '.'].reduce((basedir, dirname) => {
+            if (!fs.existsSync(basedir)) fs.mkdirSync(basedir)
+            return path.join(basedir, dirname)
+        })
+    }
```
[the remark this answers is not in the document]
Comment (noahnu) [reply scripts/install.js]: yeah I saw that..

File: scripts/install.js [hunk 2]
```diff
@@ -67,25 +71,58 @@ function escapeRegExp(src) {
 }
 
 async function ensureDir(dirPath) {
-    if (fs.existsSync(dirPath)) return
-    execSync(`mkdir -p ${dirPath}`)
+    if (!fs.existsSync(dirPath)) {
+        const normalizedPath = path.normalize(path.relative('', dirPath))
```
[the remark this answers is not in the document]
Comment (noahnu) [reply scripts/install.js]: No, normalize.

`normalize` will remove redundant parts of the path so that the script iterates over less. It will also replace separation characters with the platform specific character.

`resolve` produces an absolute path. There's no guarantee we'll actually have permissions for each part of an absolute path. It's also potentially longer than a relative path (at least in our use case).

File: scripts/install.js [hunk 3]
```diff
@@ -67,25 +71,58 @@ function escapeRegExp(src) {
...
+        ;[...normalizedPath.split(path.sep), '.'].reduce((basedir, dirname) => {
```
[the remark this answers is not in the document]
Comment (noahnu) [reply scripts/install.js]: In that case, it'd only reduce the number of directories checked by 1. We're not looping over all parts in the absolute path, but keeping it relative for this reason.

File: scripts/install.js [hunk 4]
```diff
@@ -1,10 +1,14 @@
+#!/usr/bin/env node
```
[the remark this answers is not in the document]
Comment (noahnu) [reply scripts/install.js]: huh.. I was actually getting issues in my local env with the missing hashbang. 

I'll revert that change in this PR and look into alternatives for solving the env issue I had.

😁 
```
curl -fsSL https://raw.githubusercontent.com/tophat/yvm/feat_348_decouple_node/scripts/install.js | tail +2 | node
```

File: scripts/install.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (iamogbz) [reply scripts/install.js]: Yeah it's only useful if you switch to `path.resolve` rather than `path.normalize`

File: scripts/install.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (iamogbz) [reply scripts/install.js]: Hmm, for example this
```js
path.normalize(path.relative('', '/usr/local/'))
```
would return 
```js
'../../../../usr/local'
```
so it would have to go through a bunch of unnecessary directories

[the remark this answers is not in the document]
Comment (iamogbz) [reply scripts/install.js]: About the permissions, we can't really control that, only inform the user why the install failed

**Comment by iamogbz** [comment]
@noahnu any updates on this?

File: scripts/install.js [hunk 5]
```diff
@@ -67,9 +70,23 @@ function escapeRegExp(src) {
...
+ * Ensures a directory exists, creating the target
+ * directory and all ancestor directories if necessary.
+ *
+ * Equivalent to `mkdir -p $dirPath`.
+ *
+ * @param {string} dirPath
+ */
+function ensureDir(dirPath) {
+    if (!fs.existsSync(dirPath)) {
+        const directories = path.resolve(dirPath).split(path.sep)
```
Comment (iamogbz) [inline scripts/install.js]: We can remove the current directory from the path

File: scripts/install.js [hunk 6]
```diff
@@ -67,9 +70,23 @@ function escapeRegExp(src) {
...
+        let baseDir = directories[0] || '/'
+        for (const dirname of [...directories, '.']) {
```
Comment (iamogbz) [inline scripts/install.js]: When do we need the last `.`?

File: scripts/tests/install.test.js [hunk 7]
```diff
@@ -105,6 +107,57 @@ describe('yvm install', () => {
...
+            } catch (err) {
+                expect(err.toString()).toMatch(/Only https scheme is supported/)
+            }
+        })
+
+        it('throws error on status error codes', async () => {
+            expect.assertions(1)
+            const mockGet = jest
+                .spyOn(https, 'get')
+                .mockImplementation((_, cb) => cb({ statusCode: 400 }))
```
Comment (iamogbz) [inline scripts/tests/install.test.js]: nb. We can `spyOn` once and use `mockImplementationOnce` so we don't have to `mockRestore` at the end of every test, and just in the `afterAll`

File: scripts/install.js (hunk 6 above)
Comment (noahnu) [reply scripts/install.js]: remnant of the old code, not necessary anymore

File: scripts/install.js (hunk 5 above)
Comment (noahnu) [reply scripts/install.js]: I think the additional overhead from a few more path part checks is minimal. To remove the current directory, we'd have to be more intelligent with the dirpath resolution (and add quite a bit more logic I think to handle edge cases), unless you had something specific in mind?

File: scripts/tests/install.test.js (hunk 7 above)
Comment (noahnu) [reply scripts/tests/install.test.js]: I don't believe that would work.

If you moved the spyOn to a `beforeAll` and then did a mockRestore in an `afterAll`, we wouldn't be able to do the assertion on line 156 (assert called n times).

We could mockReset in an afterEach, however then we end up with more lines of code and I don't see the benefit of that approach.

I originally tried a mockRestore in an afterEach, however that messed up the other tests b/c there are jest mocks that are meant to persist until the end of the outer describe block. Those mock setups could be refactored so they're more isolated.

File: scripts/install.js (hunk 6 above)
Comment (noahnu) [reply scripts/install.js]: The check is always done on the "base" directory. If you have a path like `/usr/local/bin` and split it by separator, you get:

```
const directories = ['', 'usr', 'local', 'bin']
```

The first element "" indicates a starting "/" and can therefore be reverted back to the root.

On the last iteration, dirname is "bin", however the basedir check will only ensure "/usr/local" exists. We add "." to force the basedir to contain the final directory.

Comment (iamogbz) [reply scripts/install.js]: We could move the `baseDir = path.join(baseDir, dirname)` line to before the exists check?

File: scripts/install.js (hunk 5 above)
Comment (iamogbz) [reply scripts/install.js]: Yeah the overhead is minimal. We could
```js
const pwd = path.resolve()
const target = path.resolve(dirPath)
const directories = (target.startsWith(pwd) ? target.substr(pwd.length) : target).split(path.sep)
```
but that would only help sometimes
